### PR TITLE
Initial import of rendering / modelling formats

### PR DESCRIPTION
**Pull request by hakanai**
Imported from [my organisation's repository](https://github.com/ephemeral-laboratories/modelling_formats).

Since this is a new directory - is "rendering" a good name for this stuff? In IANA, these formats are classified as just "model", but the word has too many different meanings in software.

About these formats:

FBX - "Filmbox" format originally by Kaydara, now owned by Autodesk, standard interchange format between modelling software and game engines and the like.

GLB - Khronos group's standard binary format for model exchange. But unfortunately a terrible format for Kaitai. I implemented it as far in as was possible anyway.

The above two file types can be created using Blender. (Actually, Blender's own format is one that I want to take a look at as well.)

PMD, PMX - Model formats used by animation software MikuMikuDance (PMX newer, PMD older), many samples can be found on bowlrowl and similar sites.

VMD - Animation data format used by MikuMikuDance.

### CODE REVIEW

File: rendering/fbx.ksy [hunk 1]
```diff
@@ -0,0 +1,148 @@
+meta:
+  id: fbx
```
Comment (KOLANICH) [inline rendering/fbx.ksy:2]: Maybe `kaydara_fbx`?

File: rendering/fbx.ksy [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+  title: FBX 3D model interchange format
```
Comment (KOLANICH) [inline rendering/fbx.ksy:3]: The word `format` is usualy redundant. Maybe just `Kaydara FiLMBOX 3D Model`?

File: rendering/fbx.ksy [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  file-extension: fbx
```
Comment (KOLANICH) [inline rendering/fbx.ksy:4]: It'd be nice to add `application: Autodesk/Kaydara FiLMBOX`

File: rendering/fbx.ksy [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    type: node_record
+    repeat: until
+    repeat-until: _.end_offset == 0
+
+types:
+
+  header:
+    seq:
+      - id: magic
+        contents: "Kaydara FBX Binary  \0"
```
Comment (KOLANICH) [inline rendering/fbx.ksy:22]:
```yaml
type: strz
valid: "Kaydara FBX Binary "
```

File: rendering/fbx.ksy [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        size: 20
+      - id: more_magic
+        contents: [0x1A, 0x00]
```
Comment (KOLANICH) [inline rendering/fbx.ksy:25]: Looks like `u2`. You can use `valid` here too.

File: rendering/fbx.ksy (hunk 2 above)
Comment (hakanai) [reply rendering/fbx.ksy:3]: I think I'll keep normal capitalisation of Filmbox but otherwise agreed.

File: rendering/fbx.ksy [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      - id: pseudo_type
+        type: u1
+      - id: element_count
+        type: u4
+    seq:
+      - id: elements
+        type:
+          switch-on: pseudo_type
+          cases:
+            1: s4
```
Comment (KOLANICH) [inline rendering/fbx.ksy:141]: The values should probably be members of a enum.

File: rendering/fbx.ksy (hunk 3 above)
Comment (hakanai) [reply rendering/fbx.ksy:4]: Will add. I had wondered about this. The native application was Filmbox but now it's used by damn near everybody. Everybody using the Autodesk FBX SDK for a start, but then even implementations in open source like Blender's.

File: rendering/glb.ksy [hunk 7]
```diff
@@ -0,0 +1,64 @@
+meta:
+  id: glb
+  title: glTF binary format
```
Comment (KOLANICH) [inline rendering/glb.ksy:3]: Nice, long-waited format :+1:

File: rendering/fbx.ksy [hunk 8]
```diff
@@ -0,0 +1,148 @@
+meta:
+  id: fbx
+  title: FBX 3D model interchange format
+  file-extension: fbx
+  encoding: UTF-8
+  endian: le
+  license: MIT
+
```
Comment (KOLANICH) [inline rendering/fbx.ksy:8]: Lacks `eoc` and `doc-ref`. Also may make sense to provide a link to a publicly-available sample file.

File: rendering/fbx.ksy (hunk 5 above)
Comment (hakanai) [reply rendering/fbx.ksy:25]: Is valid just the same as contents but you get a different exception when it's wrong?

File: rendering/glb.ksy [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  file-extension: glb
+  endian: le
+  license: MIT
+  xref:
+    doc-ref: https://github.com/KhronosGroup/glTF/tree/master/specification/2.0#binary-gltf-layout
```
Comment (KOLANICH) [inline rendering/glb.ksy:8]: Incorrect use of `doc-ref`. Feel free to merge your block with https://github.com/kaitai-io/kaitai_struct_formats/issues/209 Also add `Closes: #209` into PR text.

File: rendering/glb.ksy [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+      - id: data
+        type: str
+        encoding: UTF-8
+        size-eos: true
+        doc: |
+          This is where GLB deviates from being an elegant format.
+          To parse the rest of the file, you have to parse the JSON first.
+
+  bin:
+    seq:
```
Comment (KOLANICH) [inline rendering/glb.ksy:55]: Looks unfinished.

File: rendering/mmd_pmd.ksy [hunk 11]
```diff
@@ -0,0 +1,368 @@
+meta:
+  id: mmd_pmd
+  title: MMD model data (older format)
```
Comment (KOLANICH) [inline rendering/mmd_pmd.ksy:3]: I guess here it shouldn't be abbreviated. Also please put each format into an own PR.

File: rendering/fbx.ksy (hunk 4 above)
Comment (hakanai) [reply rendering/fbx.ksy:22]: I tried this, and I get an error that I don't fully understand.

```
/types/header/seq/0: parsing expression 'Kaydara FBX Binary  ' failed on 1:9, expected "or" | CharsWhile(Set( , n)) | "\\\n" | End
```

Wait, is `valid` expecting an _expression_?

File: rendering/mmd_pmx.ksy [hunk 12]
```diff
@@ -0,0 +1,569 @@
...
+        type: len_string
+      - id: diffuse
+        type: f4_4
+      - id: specular
+        type: f4_3
+      - id: shininess
+        type: f4
+      - id: ambient
+        type: f4_3
+      - id: flag
```
Comment (KOLANICH) [inline rendering/mmd_pmx.ksy:174]: Very ambigous name. Is it a single boolean (in this case name should start from `is_` and describe the meaning). Or is it binary flags (in this case it should be not `u1`, but a type, in which bit-sized numbers reside, `b1` is a bit and has boolean type instead)?

File: rendering/mmd_pmx.ksy [hunk 13]
```diff
@@ -0,0 +1,569 @@
...
+        type: u1
+      - id: edge_color
+        type: f4_4
+      - id: edge_size
+        type: f4
+      - id: texture_index
+        type:
+          switch-on: _root.header.texture_index_size
+          cases:
+            1: u1
```
Comment (KOLANICH) [inline rendering/mmd_pmx.ksy:184]: Again should be enums.

File: rendering/fbx.ksy (hunk 8 above)
Comment (hakanai) [reply rendering/fbx.ksy:8]: `eoc`?

File: rendering/glb.ksy (hunk 9 above)
Comment (hakanai) [reply rendering/glb.ksy:8]: Incorrect use? I originally had it one level up but the compiler gave an error that didn't tell me how to fix it, and then didn't complain when I moved it inside the refs section.

What's the right place?

I have to admit, I'm not used to working with doc-ref because I usually work on formats with zero documentation.

File: rendering/fbx.ksy (hunk 4 above)
Comment (KOLANICH) [reply rendering/fbx.ksy:22]: How about `'"Kaydara FBX Binary "'`?

File: rendering/glb.ksy (hunk 10 above)
Comment (hakanai) [reply rendering/glb.ksy:55]: In what sense? A GLB file is made up of data chunks. The chunk can be either a JSON chunk or a binary chunk. This is a binary chunk. There's no way to know what sort of data you're looking at in here though, because, for that, you'd have to first parse the JSON to get the offsets out of there.

File: rendering/fbx.ksy (hunk 5 above)
Comment (KOLANICH) [reply rendering/fbx.ksy:25]: `valid` allows multiple ways of validation, `valid: something` is just a shortcut for `valid: eq: something`

File: rendering/fbx.ksy (hunk 8 above)
Comment (KOLANICH) [reply rendering/fbx.ksy:8]: `doc`. A typo.

File: rendering/mmd_pmd.ksy (hunk 11 above)
Comment (hakanai) [reply rendering/mmd_pmd.ksy:3]: I'll just put in both like `MMD (MikuMikuDance)` since MMD is what people are going to be using as their search term.

As for splitting the PR... ugh. Yeah, sure. Once I'm done with the first round of changes I'll split it and resubmit.

File: rendering/glb.ksy (hunk 9 above)
Comment (KOLANICH) [reply rendering/glb.ksy:8]: What's the right place?

Not in `meta`, just under `doc`, wuich is a sibling of `meta`.

File: rendering/mmd_pmx.ksy (hunk 13 above)
Comment (hakanai) [reply rendering/mmd_pmx.ksy:184]: The meaning of the value is "the number of bytes in the index". Usually when you have a number of things you want to use an int, but here you're suggesting an enum.

Perhaps it might be better to keep as int but declare `valid` for the index size stuff as only valid if 1,2,4?

Comment (hakanai) [reply rendering/mmd_pmx.ksy:184]: So I will add new types for each index type like this:

```yaml
  bone_index:
    doc: Variable-length type storing a bone index
    seq:
      - id: value
        type: switch-on: _root.header.bone_index_size
        cases:
          1: u1
          2: u2
          4: u4
```

File: rendering/glb.ksy (hunk 10 above)
Comment (KOLANICH) [reply rendering/glb.ksy:55]: Please look at https://github.com/kaitai-io/kaitai_struct_formats/pull/183/files#diff-ce5a8a6f65e08325535c534e8222ecbd9f53a9749ba538b04cf857b5b07a0795R89 and https://github.com/kaitai-io/kaitai_struct_formats/pull/213/files#diff-35cf895592a530eee852a5354aaa16e650100fdfddb1e81752f7e6d0ef036786R115 

for the ideas how one can try to solve the issue of having data in some another format.

File: rendering/fbx.ksy (hunk 8 above)
Comment (hakanai) [reply rendering/fbx.ksy:8]: Ah. Well, they don't document the format so there was nothing to put in I guess. Blender had an attempt at documenting the format - is it okay to put in a third-party document describing the format?

File: rendering/glb.ksy (hunk 9 above)
Comment (hakanai) [reply rendering/glb.ksy:8]: Aha. It would have been nice for the compiler to say that, instead of just saying that it was wrong.

File: rendering/fbx.ksy (hunk 4 above)
Comment (hakanai) [reply rendering/fbx.ksy:22]: Works. Same workaround works for the array one as well.

File: rendering/fbx.ksy (hunk 5 above)
Comment (hakanai) [reply rendering/fbx.ksy:25]: Ah nice, a declarative predicate language.

File: rendering/fbx.ksy (hunk 6 above)
Comment (hakanai) [reply rendering/fbx.ksy:141]: Having issues getting this to compile with an enum... 

```
C:\Users\Trejkaz\Documents\GitHub\kaitai_struct_formats\rendering>kaitai-struct-compiler -t ruby fbx.ksy
/types/property/seq/1/type/cases/Str(i)/type: can't pass argument #0 of type Int1Type(true) into parameter `pseudo_type` of type EnumType(List(pseudo_type),Int1Type(false))
```

File: rendering/glb.ksy (hunk 10 above)
Comment (hakanai) [reply rendering/glb.ksy:55]: Neither of those appear to actually solve the issue, but rather push it out to a single-language solution, which doesn't really fit with the spirit of Kaitai anyway.

File: rendering/mmd_pmx.ksy (hunk 13 above)
Comment (hakanai) [reply rendering/mmd_pmx.ksy:184]: Now I have removed repetition in the structures above, and have instead 6 repetitive types which only differ in three places:
- their name
- the doc comment
- the field they use to switch on

File: rendering/fbx.ksy (hunk 8 above)
Comment (KOLANICH) [reply rendering/fbx.ksy:8]: They don't but you can put there some words from you, at least a brief summary of the format is needed: what its purpose is, what it is it used for, + anything you consider non-obvious.

File: rendering/glb.ksy (hunk 9 above)
Comment (KOLANICH) [reply rendering/glb.ksy:8]: > Is the compiler just not validating this section at all?

It is likely that in this section keys can be any strings.

File: rendering/glb.ksy (hunk 10 above)
Comment (KOLANICH) [reply rendering/glb.ksy:55]: >Neither of those appear to actually solve the issue, but rather push it out to a single-language solution which doesn't really fit with the spirit of Kaitai anyway.

I completely aggree.

>What would be nice is a native JSON type in Kaitai, so that we can then just traverse into the tree of JSON objects.

It would be nice to have KS extensible, because blessing JSON over other serialization formats, including custom ones, is incorrect. And the impl still will be language-dependent. The same way as runtimes are and KS itself is. All of this is just an abstraction layer around languages.

>So I guess one valid solution is, define a new format for JSON and import that format, so we can declare the type of this piece as json. And then say that implementing the JSON format is up to the reader. LOL

So a valid solution is finishing and landing interfaces proposal first, and then standardizing some interfaces and creating a standard library for each interface. See the discussion in #457.

File: rendering/fbx.ksy (hunk 6 above)
Comment (KOLANICH) [reply rendering/fbx.ksy:141]: I have encountered this some times. To solve that I had to sometimes to cast to ints and back. But sometimes passing enums just works. Needs a bit of experimentation. But if a param is a enum, it should receve a enum, if it is just wn int, it should receive an int.

File: rendering/mmd_pmx.ksy (hunk 13 above)
Comment (KOLANICH) [reply rendering/mmd_pmx.ksy:184]: You can pass the condition in `switch-on` as a param, BTW.

Comment (hakanai) [reply rendering/mmd_pmx.ksy:184]: I think I tried that in this situation once in the past and it didn't work, but it will be tried.

Comment (hakanai) [reply rendering/mmd_pmx.ksy:184]: OK, I got it down to:

```yaml
      - id: indices
        type: sized_index(_root.header.vertex_index_size)
```

I think I'll call this _good enough_ but it would be nice if I could write

```yaml
      - id: indices
        type: sized_index(index_id::vertex)
```

But (a) I'd be declaring an enum with fictional values just to use it as a param and (b) I'm 99% sure that the last time I tried using enums in params they didn't work anyway.

Comment (generalmimon) [reply rendering/mmd_pmx.ksy:184]: > But (a) I'd be declaring an enum with fictional values just to use it as a param and (b) I'm 99% sure that the last time I tried using enums in params they didn't work anyway.

They were added in the 0.9 release, see the [release notes](https://kaitai.io/news/2020/10/16/kaitai-struct-v0.9-released.html):

> - Major bugfixes:
>   - `params/type` - add support for:
>     - specific user types
>     - `enum` types (#413)
>     - byte arrays (`bytes`)
>     - arrays (`u2[]`, `struct[]`, etc.)

File: rendering/glb.ksy (hunk 10 above)
Comment (hakanai) [reply rendering/glb.ksy:55]: I wasn't saying that json would be the only format to implement, nor was I saying that json would be implemented by kaitai itself. I was just saying that we could implement a placeholder 'type' for any given unknown type, and then how that's implemented is up to the future. If the future happens to be some interface where the type is converted to something else using external code, that's fine. If someone figures out how to complete a kts file for json, that's fine too.

File: rendering/glb.ksy (hunk 9 above)
Comment (hakanai) [reply rendering/glb.ksy:8]: Sounds like a recipe for disaster...

In any case, it's fixed now.

File: rendering/fbx.ksy (hunk 8 above)
Comment (hakanai) [reply rendering/fbx.ksy:8]: Got it.

File: rendering/fbx.ksy (hunk 1 above)
Comment (hakanai) [reply rendering/fbx.ksy:2]: Should I be changing the filename to match? I notice the generated code ends up using this ID as the classname instead of the filename of the ksy. (Sucks that we have essentially the same info in two places...)

File: rendering/mmd_pmx.ksy (hunk 12 above)
Comment (hakanai) [reply rendering/mmd_pmx.ksy:174]: Similar problem with the ones below as well.

I was able to figure out that this one contains _multiple_ flags, but have no docs on what the flags are. At least for now I'll rename it to `flags`. The ones below are more interesting though.

**Comment by hakanai** [comment]
Closing as I'm splitting into multiple PRs as previously discussed.
